### PR TITLE
Add notifications 

### PR DESCRIPTION
**Pull request by MPepito**
# Add New Notification Features for Lessons

This PR introduces several notification functionalities for the lessons management system. Below are the details of the changes implemented:

### 1. **Notification: Lesson Confirmation by Tutor**
- **Trigger**: When `lesson.status` changes from `PENDING_TUTOR_CONFIRMATION` to `CONFIRMED`.
- **Description**: Sends a notification to the student informing them that the tutor has confirmed their lesson.
- **Notification Message**:
  - **Title**: "Lesson Confirmed!"
  - **Body**: "Your lesson with `<Tutor Name>` has been confirmed. See you on `<Time Slot>`!"

### 2. **Notification: New Lesson Request for Tutor**
- **Trigger**: When `lesson.status` changes from `MATCHING` to `PENDING_TUTOR_CONFIRMATION`.
- **Description**: Sends a notification to the tutor when a student requests a lesson with them.
- **Notification Message**:
  - **Title**: "New Lesson Request"
  - **Body**: "A student has requested a lesson. Please confirm or decline."

### 3. **Notification: Tutor Proposal Confirmed by Student**
- **Trigger**: When `lesson.status` changes from `STUDENT_REQUESTED` to `CONFIRMED`.
- **Description**: Sends a notification to the tutor informing them that the student has confirmed their tutoring proposal.
- **Notification Message**:
  - **Title**: "Lesson Proposal Confirmed"
  - **Body**: "The student has confirmed your tutoring proposal for `<Lesson Title>`."

### 4. **Notification: Reminder for Upcoming Lesson**
- **Trigger**: One hour before the `lesson.timeSlot` if `lesson.status` is `CONFIRMED`.
- **Description**: Sends a reminder to both the tutor and the student about their upcoming lesson.
- **Notification Message**:
  - **For Tutor**:
    - **Title**: "Upcoming Lesson Reminder"
    - **Body**: "Your lesson with `<Student Name>` starts in one hour!"
  - **For Student**:
    - **Title**: "Upcoming Lesson Reminder"
    - **Body**: "Your lesson with `<Tutor Name>` starts in one hour!"

### New Files Added:
1. **`notifications.js`**:
   - Contains helper functions for sending notifications:
     - `sendNotification(payload)`
     - `preparePayload(title, body, token)`
2. **`notificationsHandler.js`**:
   - Contains specific notification logic, such as querying Firestore for profiles and preparing personalized messages.
3. **`lessonNotifications.js`**:
   - Implements functions for each type of notification, handling the triggers and calling helper functions from `notificationsHandler.js`.

### Refactored Files:
- **`index.js`**:
  - Imports and exports notification-related functions from `lessonNotifications.js`.
  - Ensures compatibility with existing deployed functions.

## Deployment Instructions

1. Ensure all new files are in the `functions` directory.
2. Update the `index.js` file to include the following:
   ```javascript
   const { lessonStatusChanged, lessonReminder } = require("./lessonNotifications");
   exports.lessonStatusChanged = lessonStatusChanged;
   exports.lessonReminder = lessonReminder;
   ```
3. Deploy the functions using:
   ```bash
   firebase deploy --only functions
   ```

### Manual Tests Performed:
- Verified each notification is sent under the appropriate trigger conditions.
- Confirmed that no notifications are sent if the trigger conditions are not met.
- Checked edge cases, such as missing or invalid tokens.

### Future Improvements
- Add unit tests for each notification handler.
- Optimize Firestore queries for better performance.

---

### Screenshots/Logs
- **Example Notification Log**:
  ```
  Lesson status changed from MATCHING to PENDING_TUTOR_CONFIRMATION
  Notification sent successfully to <Tutor Token>

### CODE REVIEW

File: functions/index.js [hunk 1]
```diff
@@ -1,17 +1,139 @@
 const { onSchedule } = require("firebase-functions/v2/scheduler");
 const { setGlobalOptions } = require("firebase-functions/v2");
 const { initializeApp } = require("firebase-admin/app");
+initializeApp();
```
Comment (aferrante1) [inline functions/index.js:4]: Important to initializeApp at the beginning great job!

File: functions/index.js [hunk 2]
```diff
@@ -1,17 +1,139 @@
...
 const { getFirestore } = require("firebase-admin/firestore");
 const logger = require("firebase-functions/logger");
+const { onDocumentUpdated } = require("firebase-functions/v2/firestore");
```
Comment (aferrante1) [inline functions/index.js:7]: This is really good, it allows for a cleaner and slicker code. Easier to modify for future cases.

File: functions/notification.js [hunk 3]
```diff
@@ -0,0 +1,20 @@
+const admin = require("firebase-admin");
+
+module.exports = {
```
Comment (aferrante1) [inline functions/notification.js:3]: Really smart to turn this into a function so it can be reused for every notification sent!

File: functions/index.js [hunk 4]
```diff
@@ -1,17 +1,139 @@
...
+        );
+
+        const studentPayload = preparePayload(
+          "Upcoming Lesson Reminder",
+          `Your lesson with ${tutorData.firstName} starts in one hour!`,
+          studentData.token
+        );
+
+        // Send notifications
+        await Promise.all([
```
Comment (aferrante1) [inline functions/index.js:116]: Use of Promise.all for parallel data fetching and notification sending is efficient.

File: functions/index.js [hunk 5]
```diff
@@ -1,17 +1,139 @@
...
+        ]);
+
+        if (tutorQuery.empty || studentQuery.empty) {
+          console.error("Tutor or Student profile not found.");
+          return;
+        }
+
+        const tutorData = tutorQuery.docs[0].data();
+        const studentData = studentQuery.docs[0].data();
+
```
Comment (aferrante1) [inline functions/index.js:101]: Ensure that tutorData.token and studentData.token are valid before calling sendNotification. Invalid or missing tokens could throw errors.

File: functions/index.js [hunk 6]
```diff
@@ -1,17 +1,139 @@
...
+  if (after.status === "STUDENT_REQUESTED" && before.tutorUid.length < after.tutorUid.length) {
+    console.log("Tutor added to the lesson's tutorUid.");
+    await notifyStudentAsTutorOfferedToTeach(after);
+  }
+});
+
+
+exports.lessonReminder = onSchedule("every 5 minutes", async (event) => {
+  const now = new Date();
+  const oneHourFromNow = new Date(now.getTime() + 60 * 60 * 1000);
```
Comment (aferrante1) [inline functions/index.js:52]: The code follows a clear flow formatting dates, querying Firestore, processing results, and sending notifications.

File: functions/lessonNotifications.js [hunk 7]
```diff
@@ -0,0 +1,153 @@
+const { getFirestore } = require("firebase-admin/firestore");
+const { sendNotification, preparePayload } = require("./notification");
+
+
+const db = getFirestore();
+
+module.exports = {
```
Comment (aferrante1) [inline functions/lessonNotifications.js:7]: Each function has a clearly defined responsibility. This separation of concerns makes the code more readable and maintainable.

File: functions/lessonNotifications.js [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+    } catch (error) {
+      console.error("Error in notifyStudentLessonPendingReview:", error);
+    }
+  },
+
+    // Notify the student whenever an additional tutor offers to teach a lesson he requested
+  notifyStudentAsTutorOfferedToTeach: async (lesson) => {
+    const studentUid = lesson.studentUid;
+
+    try {
```
Comment (aferrante1) [inline functions/lessonNotifications.js:39]: Proper use of try-catch blocks ensures unexpected errors do not crash the system.

File: functions/lessonNotifications.js [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+        // Prepare the notification payload for the tutor
+        const tutorPayload = preparePayload(
+          "Lesson Confirmed!",
+          `Great news! The student has accepted your tutoring offer for ${lesson.title}.`,
+          tutorData.token
+        );
+
+        // Send the notification
+        await sendNotification(tutorPayload);
+        console.log("Notification sent to tutor for lesson confirmation.");
```
Comment (aferrante1) [inline functions/lessonNotifications.js:121]: Logging meaningful error messages and process updates help in debugging.

**Review by aferrante1 — APPROVED** [review]
